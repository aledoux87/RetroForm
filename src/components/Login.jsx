import React, { Component } from 'react';
import { string, func } from 'prop-types';
import UserService from '../services/userService';

class Login extends Component {
  constructor(props) {
    super(props);

    this.state = {
      login: '',
      password: '',
      email: ''
    };

    this.submitForm = this.submitForm.bind(this);
  }

  submitForm(event) {
    event.preventDefault();

    const { login, password, email } = this.state;

    console.log('<Login />', {
      login,
      password
    });

    UserService.loginUser({ login, password, email })
      .then(() => this.props.goToNextView())
      .catch(err => alert(err));

  }

  render() {

    return (
      <form className="retro-form" onSubmit={this.submitForm}>
        <h2>{this.props.title}</h2>
        {this.props.subTitle && <p>{this.props.subTitle}</p>}

        <div className="retro-field">
          <label htmlFor="login">What is your login ?</label>
          <input
            className="retro-input"
            id="login"
            name="login"
            type="text"
            defaultValue={this.state.login}
            onChange={e => this.setState({ login: e.target.value })}
          />
        </div>

        <div className="retro-field">
          <label htmlFor="password">What is your password ?</label>
          <input
            className="retro-input"
            id="password"
            name="password"
            type="password"
            defaultValue={this.state.password}
            onChange={e => this.setState({ password: e.target.value })}
          />
        </div>

        <div className="retro-field">
          <label htmlFor="email">What is your email ?</label>
          <input
            className="retro-input"
            id="email"
            name="email"
            type="email"
            defaultValue={this.state.email}
            onChange={e => this.setState({ email: e.target.value })}
          />
        </div>

        <button
          className="retro-button"
          type="submit">
          NEXT
        </button>
      </form>
    );
  }
}

Login.propTypes = {
  title: string.isRequired,
  subTitle: string,
  goToNextView: func
};

export default Login;
