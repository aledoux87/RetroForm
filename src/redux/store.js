import { combineReducers, createStore, compose, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import user from './user';

const rootReducer = combineReducers({
  user
});

export const makeStore = (state = {}) => {
  // prettier-ignore
  const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

  return createStore(rootReducer, state, composeEnhancers(applyMiddleware(thunk)));
};

const store = makeStore();

export default store;
