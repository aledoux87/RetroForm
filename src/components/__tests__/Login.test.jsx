import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import { makeStore } from '../../redux/store';
import Login from '../Login';
import UserService from '../../services/userService';

jest.mock('../../services/userService');

describe('<Login />', () => {
  const { location } = window;
  const store = makeStore();

  beforeAll(() => {
    delete window.location;
    window.location = {
      href: ''
    };
  });

  afterAll(() => {
    window.location = location;
  });

  it('Should Component Renders With Default Behaviour', () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <Login title="LoremIpsum" />
        </Router>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('Should Component Renders With Props', async () => {
    render(
      <Provider store={store}>
        <Router>
          <Login title="LoremIpsum" subTitle="Lorem ipsum dolor sit amet" />
        </Router>
      </Provider>
    );
    expect(await screen.findByText('Lorem ipsum dolor sit amet')).toBeInTheDocument();
  });

  it("Should 'onSubmit' Function Has Been Called", async () => {
    UserService.loginUser.mockResolvedValue({
      login: 'jojododo',
      passwordStrong: 8,
      email: 'jojododo@yahoo.fr'
    });

    const { container } = render(
      <Provider store={store}>
        <Router>
          <Login title="LoremIpsum" />
        </Router>
      </Provider>
    );

    const loginInput = container.querySelector('#login');
    const passwordInput = container.querySelector('#password');
    const emailInput = container.querySelector('#email');

    fireEvent.change(loginInput, { target: { value: 'jojododo' } });
    fireEvent.change(passwordInput, { target: { value: '12345678' } });
    fireEvent.change(emailInput, { target: { value: 'jojododo@yahoo.fr' } });

    const sendButton = screen.getByText('NEXT');

    fireEvent.click(sendButton);

    await waitFor(() => {
      expect(window.location.href).not.toEqual('https://www.google.com');
      // expect(history.push).toHaveBeenCalledWith('/register')
    });
  });
});
