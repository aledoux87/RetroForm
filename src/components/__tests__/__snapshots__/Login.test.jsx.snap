// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> Should Component Renders With Default Behaviour 1`] = `
<div>
  <form
    class="retro-form"
  >
    <h2>
      LoremIpsum
    </h2>
    <div
      class="retro-field"
    >
      <label
        for="login"
      >
        What is your login ?
      </label>
      <input
        class="retro-input"
        id="login"
        name="login"
        type="text"
        value=""
      />
    </div>
    <div
      class="retro-field"
    >
      <label
        for="password"
      >
        What is your password ?
      </label>
      <input
        class="retro-input"
        id="password"
        name="password"
        type="password"
        value=""
      />
    </div>
    <div
      class="retro-field"
    >
      <label
        for="email"
      >
        What is your email ?
      </label>
      <input
        class="retro-input"
        id="email"
        name="email"
        type="email"
        value=""
      />
    </div>
    <button
      class="retro-button"
      type="submit"
    >
      NEXT
    </button>
  </form>
</div>
`;
